mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala

```diff
@@ -129,17 +129,21 @@ object Bucketizer extends DefaultParamsReadable[Bucketizer] {
     if (feature == splits.last) {
       splits.length - 2
     } else {
```

Comment:
index the NaN value as {splits.leng -2}
all the changes below this line are just for code refactoring.